### PR TITLE
New Resource: `azurerm_mobile_network_slice`; New Data Source: `azurerm_mobile_network_slice`

### PR DESCRIPTION
**Pull request by ziyeqf**
an overall list is below. Detail of this RP could be found [here](https://learn.microsoft.com/en-us/azure/private-5g-core/)

Though these PRs may have conflicts in merge sequence, but I think it should not block the review progress...

Over view
---
```
order | name 
order | name 
0     | azurerm_mobile_network                              ✅
1     | azurerm_mobile_network_site                         #20334
1     | azurerm_mobile_network_data_network                 #20338
1     | azurerm_mobile_network_service                      #20337
1     | azurerm_mobile_network_slice                        <- this one
1     | azurerm_mobile_network_sim_group                    #20339
2     | azurerm_mobile_network_packet_core_control_plane    
2     | azurerm_mobile_network_attached_data_network        
2     | azurerm_mobile_network_sim_policy                   
3     | azurerm_mobile_network_packet_core_data_plane       
4     | azurerm_mobile_network_attached_data_network        
5     | azurerm_mobile_network_sim        
```

Test
---
```
terraform-provider-azurerm ❯ TF_ACC=1 go test -v ./internal/services/mobilenetwork -run=TestAccMobileNetworkSlice -timeout=600m                                                     [ tengzh/mn_slice][$!][🐹 v1.19.5][⏱ 1s][10:34:09]
=== RUN   TestAccMobileNetworkSliceDataSource_complete
=== PAUSE TestAccMobileNetworkSliceDataSource_complete
=== RUN   TestAccMobileNetworkSlice_basic
=== PAUSE TestAccMobileNetworkSlice_basic
=== RUN   TestAccMobileNetworkSlice_requiresImport
=== PAUSE TestAccMobileNetworkSlice_requiresImport
=== RUN   TestAccMobileNetworkSlice_complete
=== PAUSE TestAccMobileNetworkSlice_complete
=== RUN   TestAccMobileNetworkSlice_update
=== PAUSE TestAccMobileNetworkSlice_update
=== CONT  TestAccMobileNetworkSliceDataSource_complete
=== CONT  TestAccMobileNetworkSlice_complete
=== CONT  TestAccMobileNetworkSlice_requiresImport
=== CONT  TestAccMobileNetworkSlice_basic
=== CONT  TestAccMobileNetworkSlice_update
--- PASS: TestAccMobileNetworkSliceDataSource_complete (471.16s)
--- PASS: TestAccMobileNetworkSlice_complete (496.23s)
--- PASS: TestAccMobileNetworkSlice_basic (497.84s)
--- PASS: TestAccMobileNetworkSlice_requiresImport (533.85s)
--- PASS: TestAccMobileNetworkSlice_update (678.82s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/mobilenetwork 680.034s

```

### CODE REVIEW

File: internal/services/mobilenetwork/helpers.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			return resp, strconv.Itoa(resp.StatusCode), nil
+		},
+		MinTimeout:                10 * time.Second,
+		PollInterval:              10 * time.Second,
+		ContinuousTargetOccurence: 6,
+		Timeout:                   time.Until(deadline),
+	}
+
+	if _, err := stateConf.WaitForStateContext(ctx); err != nil {
+		return fmt.Errorf("waiting for deleting %s: %+v", id, err)
```
Comment (catriona-m) [inline internal/services/mobilenetwork/helpers.go:40]:
```suggestion
		return fmt.Errorf("waiting for deletion of %s: %+v", id, err)
```

File: internal/services/mobilenetwork/mobile_network_slice_data_source.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"mobile_network_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: mobilenetwork.ValidateMobileNetworkID,
+		},
```
Comment (catriona-m) [inline internal/services/mobilenetwork/mobile_network_slice_data_source.go:48]: We can remove the ForceNews here given that this is the datasource

File: internal/services/mobilenetwork/mobile_network_slice_data_source.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			client := metadata.Client.MobileNetwork.SliceClient
+			mobileNetworkId, err := mobilenetwork.ParseMobileNetworkID(metaModel.MobileNetworkMobileNetworkId)
+			if err != nil {
+				return err
+			}
+
+			id := slice.NewSliceID(mobileNetworkId.SubscriptionId, mobileNetworkId.ResourceGroupName, mobileNetworkId.MobileNetworkName, metaModel.Name)
```
Comment (catriona-m) [inline internal/services/mobilenetwork/mobile_network_slice_data_source.go:100]: Given that we are using the `mobile_network_id` and `name` here to build the SliceID, would it make sense to just have the datasource require just the SliceID to begin with? Then the MobileNetworkID could be parsed from that and we would only need one argument.

File: internal/services/mobilenetwork/mobile_network_slice_resource.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+			}
+
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			properties := resp.Model
+			if properties == nil {
+				return fmt.Errorf("retrieving %s: properties was nil", id)
```
Comment (catriona-m) [inline internal/services/mobilenetwork/mobile_network_slice_resource.go:176]:
```suggestion
				return fmt.Errorf("retrieving %s: properties were nil", id)
```

File: website/docs/d/mobile_network_slice.html.markdown [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name which should be used for this Mobile Network Slice. Changing this forces a new Mobile Network Slice to be created.
+
+* `mobile_network_id` - (Required) The ID of Mobile Network which the Mobile Network Slice belongs to. Changing this forces a new Mobile Network Slice to be created.
```
Comment (catriona-m) [inline website/docs/d/mobile_network_slice.html.markdown:33]: We can omit the force new warnings here given that this is the datasource

```suggestion
* `name` - (Required) Specifies the name which should be used for this Mobile Network Slice.

* `mobile_network_id` - (Required) The ID of Mobile Network which the Mobile Network Slice belongs to. 
```

File: website/docs/d/mobile_network_slice.html.markdown [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+## Attributes Reference
+
+In addition to the Arguments listed above - the following Attributes are exported:
+
+* `id` - The ID of the Mobile Network Slice.
+
+* `location` - The Azure Region where the Mobile Network Slice should exist. Changing this forces a new Mobile Network Slice to be created.
```
Comment (catriona-m) [inline website/docs/d/mobile_network_slice.html.markdown:41]:
```suggestion
* `location` - The Azure Region where the Mobile Network Slice should exist. 
```

File: website/docs/d/mobile_network_slice.html.markdown [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+* `single_network_slice_selection_assistance_information` - A `single_network_slice_selection_assistance_information` block as defined below. Single-network slice selection assistance information (S-NSSAI). Unique at the scope of a mobile network.
+
+* `description` - An optional description for this network slice.
```
Comment (catriona-m) [inline website/docs/d/mobile_network_slice.html.markdown:45]:
```suggestion
* `description` - A description of this Mobile Network Slice.
```

File: website/docs/d/mobile_network_slice.html.markdown [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+* `tags` - A mapping of tags which should be assigned to the Mobile Network Slice.
```
Comment (catriona-m) [inline website/docs/d/mobile_network_slice.html.markdown:47]:
```suggestion
* `tags` - A mapping of tags which are assigned to the Mobile Network Slice.
```

File: website/docs/r/mobile_network_slice.html.markdown [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+* `name` - (Required) Specifies the name which should be used for this Mobile Network Slice. Changing this forces a new Mobile Network Slice to be created.
+
+* `mobile_network_id` - (Required) The ID of Mobile Network which the Mobile Network Slice belongs to. Changing this forces a new Mobile Network Slice to be created.
+
+* `location` - (Required) Specifies the Azure Region where the Mobile Network Slice should exist. Changing this forces a new Mobile Network Slice to be created.
+
+* `single_network_slice_selection_assistance_information` - (Required) A `single_network_slice_selection_assistance_information` block as defined below. Single-network slice selection assistance information (S-NSSAI). Unique at the scope of a mobile network.
+
+* `description` - (Optional) An optional description for this network slice.
```
Comment (catriona-m) [inline website/docs/r/mobile_network_slice.html.markdown:59]:
```suggestion
* `description` - (Optional) A description for this Mobile Network Slice.
```

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks @ziyeqf - this is looking good so far. I have left a couple of comments inline but can take another look once those are addressed. Thanks!

File: internal/services/mobilenetwork/mobile_network_slice_data_source.go (hunk 3 above)
Comment (ziyeqf) [reply internal/services/mobilenetwork/mobile_network_slice_data_source.go:100]: My thought is `slice` is a nested resource inside `mobile network` like other resources inside `resource group`...

And with this, if it's better to require `resource_group_name`, `mobile_network_name` and `slice_name`?

File: internal/services/mobilenetwork/mobile_network_slice_resource.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+			existing, err := client.Get(ctx, id)
+			if err != nil && !response.WasNotFound(existing.HttpResponse) {
+				return fmt.Errorf("checking for existing %s: %+v", id, err)
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			properties := &slice.Slice{
```
Comment (catriona-m) [inline internal/services/mobilenetwork/mobile_network_slice_resource.go:131]:
```suggestion
			properties := slice.Slice{
```

File: internal/services/mobilenetwork/mobile_network_slice_resource.go [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+				Tags:       &model.Tags,
+			}
+
+			if model.Description != "" {
+				properties.Properties.Description = &model.Description
+			}
+
+			properties.Properties.Snssai = expandSnssaiModel(model.Snssai)
+
+			if err := client.CreateOrUpdateThenPoll(ctx, id, *properties); err != nil {
```
Comment (catriona-m) [inline internal/services/mobilenetwork/mobile_network_slice_resource.go:143]:
```suggestion
			if err := client.CreateOrUpdateThenPoll(ctx, id, properties); err != nil {
```

File: internal/services/mobilenetwork/mobile_network_slice_resource.go [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			properties := resp.Model
```
Comment (catriona-m) [inline internal/services/mobilenetwork/mobile_network_slice_resource.go:174]: Could we change this the same way as suggested here? https://github.com/hashicorp/terraform-provider-azurerm/pull/20339#discussion_r1102615556

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks @ziyeqf I spotted a couple of other things looking through this again, but overall this is looking pretty good. It looks like some of these may apply to the other PRs for mobilenetwork too. Thanks for your work on this, hopefully we will get these all merged soon!

**Review by catriona-m — APPROVED** [review]
Thanks @ziyeqf LGTM!

**Pull request merged**